### PR TITLE
resolvectl: Add show-multicast verb to show discovered LLMNR/mDNS hosts

### LINKED ISSUE #14796: resolvectl: Add a way to show resolved's cache entries
**Is your feature request related to a problem? Please describe.**

resolvectl doesn't seem to provide a command to show a list of discovered hosts on the local network (cache entries).

**Describe the solution you'd like**

Add a command to resolvectl that lists the entries in the resolved cache.

From a quick look at resolved's code, it seems we'd need to add a dbus call to retrieve the entries from resolved's cache which would take a few filtering options to allow filtering the output (just like the `ResolveHostname` dbus call). In the dbus call we'd simply iterate over the necessary dns scope caches and construct a response.

What I'm not so sure about is how we'd best expose this in resolvectl's CLI. Two possible options that I can see:

1. Add a table with cache entries to the status output produced by resolvectl's `mdns`, `llmnr` and `dns` commands.
2. Add a explicit command (`resolvetctl discover` (like avahi-discover), `resolvectl dump`, ...) that lists all entries from all caches and allow filtering with the existing `-i`, `-p`, `-4`, `-6`, ... options.

I'm leaning towards the second option because it provides greater flexibility but I would be fine with either one.

Any opinions? This seems doable to implement on my own so I can have a go at it but it would be great to know if something like this would be accepted.

### PR DESCRIPTION
**Pull request by daandemeyer**
resolvectl currently lacks an easy way to list what hosts have been discovered on the local network via LLMNR or mDNS. With the show-multicast verb, this information is displayed (formatted in a table).

Fixes #14796

Depends on #18354

Test plan:

```
mkosi qemu
systemctl start systemd-resolved
resolvectl show-multicast
```

Output:

```
[root@archlinux ~]# resolvectl show-multicast
HOSTNAME  ADDRESS         SOURCE    
archlinux 169.254.194.202 LLMNR/IPv4
archlinux 192.168.91.17   LLMNR/IPv4
```

### CODE REVIEW

**Comment by bluca** [comment]
> Discover() returns an array of hostnames/addresses discovered via
LLMNR or Multicast DNS. It does not trigger any discover on its own

From the UI point of view, this sounds a bit confusing to me - is there a different verb that could best describe what it does? show-discovered maybe?

**Comment by boucman** [comment]
list-known maybe ?

File: src/resolve/resolvectl.c [hunk 1]
```diff
@@ -2627,14 +2712,13 @@ static int native_help(void) {
...
                "     --legend=BOOL             Print headers and additional info (default: yes)\n"
-               "\nSee the %s for details.\n"
-               , program_invocation_short_name
-               , ansi_highlight()
-               , ansi_normal()
-               , ansi_highlight()
-               , ansi_normal()
-               , link
-        );
+               "\nSee the %s for details.\n",
```
Comment (Hendrikto) [inline src/resolve/resolvectl.c]: comma placement is unrelated to the change

**Comment by Hendrikto** [comment]
I suggest `list-hosts`.

**Comment by DaanDeMeyer** [comment]
As Visual Studio Code now supports only formatting modified lines and we have a .clang-format config in the repository, I tried using clang-format while writing this PR, formatting only the modified lines. For tables, clang-format unfortunately doesn't do the right thing, so I added directives to turn off clang-format for the tables I touched to the source code.

Let me know if we want to allow these directives in the source code to allow more seamless usage of clang-format or if these should be removed. The comma change is also done by clang-format as the comma's are part of the `printf` function call.

**Comment by boucman** [comment]
list-discovered is shorter, and in the context of resolvectl it's probably clear enough...

**Comment by DaanDeMeyer** [comment]
Renamed to list-discovered (D-Bus method to ListDiscovered)

**Comment by DaanDeMeyer** [comment]
Two interesting things I noticed while testing this:

1. There seem to be no LLMNR entries in the cache when booting the mkosi container using systemd-nspawn. Anyone got any ideas why LLMNR discovery would differ depending on whether we run in a container or a VM?
2. When testing using QEMU, resolvectl will only ever contain IPv4 or IPv6 addresses for the host machine, but never both at the same time. Is this inherent to the LLMNR protocol?

File: mkosi.postinst [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+set -e
+
+mkdir -p /root/build
+echo "directory /root/build" > ~/.gdbinit
```
Comment (keszybz) [inline mkosi.postinst:5]: This assumes that the mkosi image will only be used for systemd debugging. It seems a bit presumptive, but is probably correct almost always.

File: man/org.freedesktop.resolve1.xml [hunk 3]
```diff
@@ -115,6 +115,7 @@ node /org/freedesktop/resolve1 {
       ResetStatistics();
       FlushCaches();
       ResetServerFeatures();
+      ListDiscovered(out a(stiiay) UNNAMED);
```
Comment (keszybz) [inline man/org.freedesktop.resolve1.xml]: Shouldn't the method name somehow indicate that this is about multicast dns? If we had some common name for LLMNR + mDNS, it'd be nice to use it here.

File: src/resolve/resolvectl.c (hunk 1 above)
Comment (keszybz) [reply src/resolve/resolvectl.c]: Yeah, I think this change should be removed. Those commas are intentionally and consistently placed like that to make it easier to edit the multiline string. If we were to decide that we don't want that and want to move the commas, this should be done for all the codebase.

**Review by keszybz — COMMENTED** [review]
Looks all good, except for the name. I know naming is hard, but "discovered" is pretty much meaningless without explanation. In particular, classis DNS names we have in the cache have also been "discovered", and it wouldn't be clear why they are not included in this output. We call mDNS + LLMNR "multicast dns", but that name is very close to MulticastDNS/mDNS, so can't be used directly. My other idea would be something like "local hosts", but obviously that is too close to "localhost". But I think we need to find some word that emphasizes that this about non-unicast-dns names. Maybe "list-multicast-names" or "list-multicast-hosts"?

**Comment by keszybz** [comment]
> Maybe "list-multicast-names" or "list-multicast-hosts"?

Another approach. We currently don't have a `show-cache` verb, but I'm pretty sure we will want to add this at some point. We could "attach" the functionality in this PR to that verb, by implementing `show-cache --multicast`. `--multicast` would be required or implicit for now. Later on, if we implement dumping of other entry types, we can add more switches.

I don't think cache dumping should be totally unprivileged. It may create an information leak. It seems OK for mDNS/LLMNR names, but for classic DNS we should probably go through polkit.

**Comment by DaanDeMeyer** [comment]
What about show-multicast? I agree that this is pretty narrow and maybe a larger dump-cache verb would be better but is dumping the entire dns cache really going to be useful for users? I'd think users would primarily be interested in the multicast hosts. For developers, having the dns cache could be useful but it's already possible to dump the cache by sending a signal to resolved. That should be sufficient for development purposes no?

**Comment by keszybz** [comment]
`show-multicast` is fine by me.

File: src/resolve/resolvectl.c (hunk 1 above)
Comment (daandemeyer) [reply src/resolve/resolvectl.c]: I'll see about making a PR for the entire codebase.

File: man/org.freedesktop.resolve1.xml (hunk 3 above)
Comment (daandemeyer) [reply man/org.freedesktop.resolve1.xml]: Renamed to GetMulticastHosts()

**Comment by DaanDeMeyer** [comment]
Renamed the verb to show-multicast and the D-Bus method to GetMulticastHosts().

File: src/resolve/resolvectl.c [hunk 4]
```diff
@@ -2500,6 +2500,90 @@ static int verb_log_level(int argc, char *argv[], void *userdata) {
...
+                        return log_error_errno(r, "Failed to print address: %m");
+
+                r = table_add_many(
+                        table,
+                        TABLE_STRING,
+                        canonical,
+                        TABLE_STRING,
+                        pretty,
+                        TABLE_STRING,
+                        resolve_flags_to_string((flags)));
```
Comment (yuwata) [inline src/resolve/resolvectl.c]: Redundant `()`.

File: src/resolve/resolvectl.c [hunk 5]
```diff
@@ -2500,6 +2500,90 @@ static int verb_log_level(int argc, char *argv[], void *userdata) {
...
+                if (r < 0)
+                        return r;
```
Comment (yuwata) [inline src/resolve/resolvectl.c]: Missing `log_table_xxx()`.

**Review by yuwata — CHANGES_REQUESTED** [review]

**Comment by DaanDeMeyer** [comment]
@keszybz @yuwata Thanks for the reviews. Fixed all the comments I think. PTAL.

**Comment by poettering** [comment]
hmmpf, i am not sure i like this concept. the thing is that people might believe the shown list is comprehensive, but it really is not. it's just what is cached, i.e. what was recently asked for or recently announced. there might be a lot more on the network, and there's no guarantee whatsoever that this is complete.

there's also the problem that dbus sucks for passing along huge amounts of data, and my guess is that a fully populated cache might hit the dbus message size limits if you have plenty of interfaces. 

quite frankly, i#d drop this again the way this is.

what we can add is a "dump-cache" verb or so, (or "show-cache") but it should be able to dump all kinds of RRs and then hook this up with --type= and --class= to do filtering, as well as --interface= and --protocol= to limit things. You could then do "resolvectl dump-cache --protocol=mdns --type=A --type=AAAA" to get the data this command currently shows. 

And i figure to make this work via dbus one approach could be to serialize the RR data into JSON, and then pass a memfd over to it.

sorry i didn't review this earlier. I know I am a bit late.

But i don't think we should leave this in as it is right now.

**Comment by poettering** [comment]
my key point really is: it needs to have "cache" in the name, so that people understand this is just cached info and not comprehensive. even if we wanted to neither llmnr nor mdns actually provide a way to get a complete dump of hosts. they simply don't describe a query like that.

**Comment by DaanDeMeyer** [comment]
I'm just wondering if show-cache isn't too extensive. I think this stuff is 20% of the work for 80% of the benefit. If we exclude DNS, are we really expecting so many entries that we'll reach the D-Bus message limit? I don't mind changing it to use a memory fd but I was thinking that'd be overkill. I really want to keep it to a single verb though without having to pass tons of options just to get the LLMNR/mDNS hosts that resolved knows out. I don't mind renaming to `show-multicast-cache` if that's better.

**Comment by poettering** [comment]
well, people run systems with thousands of interface, so yeah, i think this is a real problem.

I'd be fine to add a shortcut like "-M" to mean the same as "--type=a --type=aaaa --protocol=mdns --protocol=llmnr" or so, then you could do "resolvectl show-cache -M", which should be shorter than your current verb

**Comment by DaanDeMeyer** [comment]
I was assuming we'd hit resolved's cache size limit long before we'd ever run into problems. Or does that not apply to LLMNR/mDNS?

But OK, I'll implement the full cache dump version when I get some time

**Pull request merged**